### PR TITLE
(#142) add database storage information to deployment report

### PR DESCRIPTION
**Pull request by cuddlehub**

### CODE REVIEW

File: deployment_report/model/viya_deployment_report.py [hunk 1]
```diff
@@ -1,6 +1,4 @@
 ####################################################################
-# ### viya_deployment_report.py                                  ###
```
Comment (kevinlinglesas) [inline deployment_report/model/viya_deployment_report.py:2]: Why were these line removed, disrupting the existing file header.

**Review by kevinlinglesas — APPROVED** [review]
Looks good to me, but I'd like @sasjowood review to be done before merging.

File: deployment_report/model/utils/config_util.py [hunk 2]
```diff
@@ -29,22 +30,40 @@
...
+
+# database name key values
+_DBNAME_POSTGRES_ = "postgres"
+_DBNAME_CONFIG_POSTGRES_ = "sas-postgres-config"
+_DBNAME_CPSPOSTGRES_ = "cpspostgres"
+_DBNAME_CONFIG_CPSPOSTGRES_ = "sas-planning-cpspostgres-config"
 
 
-def get_cadence_version(config_maps: Dict) -> Optional[Text]:
+def get_cadence_version(resource: Dict) -> Optional[Text]:
```
Comment (sasjowood) [inline deployment_report/model/utils/config_util.py:66]: The param name should be adjusted to `resource_cache` to align with other util methods. The "param" and "return" flags should also be updated to note this change.

Example:

`:param resource_cache: The dictionary of resources gathered from the k8s deployment.`

File: deployment_report/model/utils/config_util.py [hunk 3]
```diff
@@ -29,22 +30,40 @@
...
-_HOST_KEY_ = "host"
-_NAME_KEY_ = "name"
-_PORT_KEY_ = "port"
-_TYPE_KEY_ = "type"
-_USER_KEY_ = "user"
+_HOST_KEY_ = "Host"
+_NAME_KEY_ = "Name"
+_PORT_KEY_ = "Port"
+_TYPE_KEY_ = "Type"
+_SSLENABLED_KEY_ = "SSL_Enabled"
```
Comment (sasjowood) [inline deployment_report/model/utils/config_util.py:49]: Keys in the data dictionary are camel-cased by convention, formatting for display should be done in the jinja template.

File: deployment_report/model/utils/config_util.py [hunk 4]
```diff
@@ -54,66 +73,175 @@ def get_cadence_version(config_maps: Dict) -> Optional[Text]:
...
 
+        # expected value wasn't found
+        return None
 
-def get_db_info(config_maps: Dict) -> Dict:
+    except KeyError:
+        return None
+
+
+def get_db_info(resource: Dict) -> Dict:
```
Comment (sasjowood) [inline deployment_report/model/utils/config_util.py:110]: The param name should be adjusted to `resource_cache` to align with other util methods. The "param" and "return" flags should also be updated to note this change.

Example:

`:param resource_cache: The dictionary of resources gathered from the k8s deployment.`

File: deployment_report/model/utils/config_util.py [hunk 5]
```diff
@@ -54,66 +73,175 @@ def get_cadence_version(config_maps: Dict) -> Optional[Text]:
...
     """
     Returns the db information of the targeted SAS deployment.
 
-    :param config_maps: The ConfigMap resources to evaluate.
+    :param resource: The api resource data.
+
+    :return: A dict representing the db information of the targeted SAS deployment, or None if
+             no resource information
```
Comment (sasjowood) [inline deployment_report/model/utils/config_util.py:117]: I think this sentence needs to be completed, something like "...if no resource information was cached."

File: deployment_report/model/utils/config_util.py [hunk 6]
```diff
@@ -54,66 +73,175 @@ def get_cadence_version(config_maps: Dict) -> Optional[Text]:
...
+            db_dict[db_name] = dbs
+
+    return db_dict
+
+
+def _get_db_info_v2(pgclusters: Dict) -> Dict:
+    """
+    Returns the db information of the targeted SAS deployment.
+
+    :param pgclusters: The pgclusters resource to evaluate. (RDCHANGE-434)
```
Comment (sasjowood) [inline deployment_report/model/utils/config_util.py:204]: This might be a reference to an internal Jira ticket. We should remove references to internal tickets from public repos.

File: deployment_report/model/utils/test/test_config_util.py [hunk 7]
```diff
@@ -32,11 +32,8 @@ def test_get_cadence_version(no_ingress_simulation_fixture: conftest.DSA):
     # get the resource cache
     resource_cache: Dict = no_ingress_simulation_fixture.resource_cache()
 
-    # get the ConfigMaps
-    config_maps: Dict = resource_cache[ResourceTypeValues.K8S_CORE_CONFIG_MAPS][ITEMS_KEY]
-
     # test the util method
-    assert config_util.get_cadence_version(config_maps=config_maps) == KubectlTest.Values.CADENCEINFO
+    assert config_util.get_cadence_version(resource=resource_cache) == KubectlTest.Values.CADENCEINFO
```
Comment (sasjowood) [inline deployment_report/model/utils/test/test_config_util.py:36]: Refactoring the param name (suggested in the config_util.py changes) to `resource_cache` should change this call.

File: deployment_report/model/viya_deployment_report.py [hunk 8]
```diff
@@ -243,14 +243,11 @@ def gather_details(self, kubectl: KubectlInterface,
                                                            sas_custom_resource_types=sas_custom_resource_types)
 
         #######################################################################
-        # Get configuration values from ConfigMaps                            #
+        # Create a cadence/db information                                     #
         #######################################################################
-        # get the ConfigMap resources that were cached above
-        config_maps: Dict = resource_cache[ResourceTypeValues.K8S_CORE_CONFIG_MAPS][ITEMS_KEY]
+        cadence_info: Optional[Text] = config_util.get_cadence_version(resource_cache)
```
Comment (sasjowood) [inline deployment_report/model/viya_deployment_report.py:248]: Best practice is to reference the param by name:

```python
config_util.get_cadence_version(resource_cache=resource_cache)
```

That way, if the signature of that method were to change in the future we would not need to adjust the order of params or miss the change and produce unexpected results.

File: deployment_report/model/viya_deployment_report.py [hunk 9]
```diff
@@ -243,14 +243,11 @@ def gather_details(self, kubectl: KubectlInterface,
...
 
-        # get the report values from ConfigMap resources
-        cadence_info: Optional[Text] = config_util.get_cadence_version(config_maps=config_maps)
-        db_dict: Optional[Dict] = config_util.get_db_info(config_maps=config_maps)
+        db_dict: Optional[Dict] = config_util.get_db_info(resource_cache)
```
Comment (sasjowood) [inline deployment_report/model/viya_deployment_report.py:250]: Same thing as above regarding referencing the param by name.

File: viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+                "sas.com/default-database": "true"
+            },
+            "creationTimestamp": "2021-08-03T18:48:53Z",
+            "generation": 1,
+            "labels": {
+                "sas.com/admin": "cluster-local",
+                "sas.com/deployment": "sas-viya"
+            },
+            "managedFields": [
```
Comment (sasjowood) [inline viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json:16]: The `managedFields` key can be deleted to reduce file size.

File: viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
...-cmm2t44t88\"}},{\"configMapRef\":{\"name\":\"sas-certframe-user-config-c4ch2c59m7\"}},{\"configMapRef\":{\"name\":\"sas-certframe-config-2ch97fd95b\"}}],\"image\":\"test.sas.com/test/sas-certframe:3.17.5-20210729.1627580385314\",\"name\":\"sas-certframe\",\"resources\":{\"limits\":{\"cpu\":\"500m\",\"memory\":\"500Mi\"},\"requests\":{\"cpu\":\"50m\",\"memory\":\"50Mi\"}},\"securityContext\":{\"allowPrivilegeEscalation\":false,\"capabilities\":{\"drop\":[\"ALL\"]},\"privileged\":false,\"readOnlyRootFilesystem\":true},\"volumeMounts\":[{\"mountPath\":\"/certframe-token\",\"name\":\"certframe-token\"},{\"mountPath\":\"/security\",\"name\":\"security\"},{\"mountPath\":\"/customer-provided-ca-certificates\",\"name\":\"customer-provided-ca-certificates\"}]}],\"volumes\":[{\"name\":\"certframe-token\",\"secret\":{\"secretName\":\"sas-certframe-token\"}},{\"emptyDir\":{},\"name\":\"security\"},{\"configMap\":{\"name\":\"sas-customer-provided-ca-certificates-km44fc867t\"},\"name\":\"customer-provided-ca-certificates\"}]},\"connection\":{\"host\":\"testm1.openstack.sas.com\",\"port\":5432,\"ssl\":true},\"database\":\"SharedServices\",\"internal\":false,\"rolesecret\":\"postgres-user\"}}\n",
```
Comment (sasjowood) [inline viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json:7]: This annotation can be deleted to reduce file size.

File: viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+                            "f:shutdown": {}
+                        }
+                    },
+                    "manager": "kubectl",
+                    "operation": "Update",
+                    "time": "2021-08-03T18:48:53Z"
+                }
+            ],
+            "name": "postgres",
+            "namespace": "d36363",
```
Comment (sasjowood) [inline viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json:61]: Change namespace to "test" for consistency.

File: viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+                    "apiVersion": "orchestration.sas.com/v1alpha1",
+                    "blockOwnerDeletion": true,
+                    "controller": true,
+                    "kind": "SASDeployment",
+                    "name": "sas-viya",
+                    "uid": "74904415-de32-497f-9d5e-1b06c7058305"
+                }
+            ],
+            "resourceVersion": "287826",
+            "selfLink": "/apis/webinfdsvr.sas.com/v1/namespaces/d36363/pgclusters/postgres",
```
Comment (sasjowood) [inline viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json:73]: Replace namespace value with "test" for consistency.

File: deployment_report/model/utils/config_util.py [hunk 14]
```diff
@@ -54,66 +73,175 @@ def get_cadence_version(config_maps: Dict) -> Optional[Text]:
...
+    Returns the db information of the targeted SAS deployment.
+
+    :param config_maps: The ConfigMap resources to evaluate. This is before data server operator change.
 
     :return: A dict representing the db information of the targeted SAS deployment, or None if the given
              ConfigMap does not contain the requested information.
     """
-    # look for the necessary ConfigMap
+    # initialize the return value
+    db_dict: Optional[Dict] = dict()
```
Comment (sasjowood) [inline deployment_report/model/utils/config_util.py:153]: I'm not sure this should be hinted as "Optional". It looks like this value is set to an empty dict and nothing would revert it to None before it's returned, which aligns with the return type hint for the method.

File: deployment_report/model/utils/config_util.py [hunk 15]
```diff
@@ -54,66 +73,175 @@ def get_cadence_version(config_maps: Dict) -> Optional[Text]:
...
+
     for config_map_name, config_map_details in config_maps.items():
+        dbs: Optional[Dict] = dict()
```
Comment (sasjowood) [inline deployment_report/model/utils/config_util.py:156]: Same thing as above regarding the Optional hint.

File: deployment_report/model/utils/config_util.py [hunk 16]
```diff
@@ -54,66 +73,175 @@ def get_cadence_version(config_maps: Dict) -> Optional[Text]:
...
+
+def _get_db_info_v2(pgclusters: Dict) -> Dict:
+    """
+    Returns the db information of the targeted SAS deployment.
+
+    :param pgclusters: The pgclusters resource to evaluate. (RDCHANGE-434)
+    :return: A list of dictionary representing the db information of the targeted SAS deployment.
+    """
+    # initialize the return value
+    db_dict: Optional[Dict] = dict()
```
Comment (sasjowood) [inline deployment_report/model/utils/config_util.py:208]: Other instance of an Optional hint but it doesn't look like this value can revert to a NoneType.

File: deployment_report/model/utils/config_util.py [hunk 17]
```diff
@@ -54,66 +73,175 @@ def get_cadence_version(config_maps: Dict) -> Optional[Text]:
...
+
+    for key in pgclusters:
+        try:
+            resource_definition = pgclusters[key][Keys.ResourceDetails.RESOURCE_DEFINITION]
+
+            db_name: Optional[Text] = resource_definition.get_name()
+            db_data: Optional[Dict] = resource_definition.get_spec()
+
+            dbs: Optional[Dict] = dict()
```
Comment (sasjowood) [inline deployment_report/model/utils/config_util.py:217]: Other instance of an Optional hint but it doesn't look like this value can revert to a NoneType.

File: deployment_report/templates/viya_deployment_report.html.j2 [hunk 18]
```diff
@@ -130,6 +130,26 @@
...
+  {% set kind_name = "Storage" %}
+  <div class="jq-accordion">
+    <h2>Storage</h2>
+    <div>
+      {% for db, info in report_data.kubernetes.dbInfo.items() %}
+      <h7>Database({{db}})</h7>
+      <table>
+      {% for key, value in info.items() %}
+        <tr>
+          <th>{{key}}</th>
```
Comment (sasjowood) [inline deployment_report/templates/viya_deployment_report.html.j2:143]: You can convert keys to displayable values here with: `{{ key | title }}`.

Also, in general, variables should be formatted like `{{ <var_name> }}` (i.e., a space between the var_name and the brackets) for consistency with the rest of the file.

**Review by sasjowood — CHANGES_REQUESTED** [review]

File: deployment_report/model/utils/config_util.py (hunk 3 above)
Comment (cuddlehub) [reply deployment_report/model/utils/config_util.py]: Add new class DatabaseDetails to viya_deployment_report_keys.py.

File: viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json (hunk 11 above)
Comment (cuddlehub) [reply viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json]: Removed.

File: viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json (hunk 12 above)
Comment (cuddlehub) [reply viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json]: updated.

File: viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json (hunk 13 above)
Comment (cuddlehub) [reply viya_ark_library/k8s/test_impl/response_data/resources_pgclusters.webinfdsvr.sas.com.json]: updated.

File: deployment_report/model/static/viya_deployment_report_keys.py [hunk 19]
```diff
@@ -78,3 +78,15 @@ class Relationship(object):
...
+        """
+        Class defining static references to keys in Database details dictionaries in resulting
+        ViyaDeploymentReport data structure.
+        """
+        DBHOST = "Host"
+        DBNAME = "Name"
+        DBPORT = "Port"
+        DBTYPE = "Type"
+        DBSSL = "SSL_Enabled"
+        DBCONN = "Connection"
```
Comment (sasjowood) [inline deployment_report/model/static/viya_deployment_report_keys.py:92]: Moving the keys into this file was a good idea. Thank you for doing so.

Unfortunately, the way they're defined here still yields the same issue as before. If you take a look at the other keys in this file, everything is in lower camel-case (e.g., "resourceType") and none of the keys contain an underscore (other than a few exceptions that start with an underscore to represent metadata about the report). The DatabaseDetails keys are mostly title-cased (i.e., first letter capitalized). So in the resulting json most keys will be defined like:

```json
{
    "resourceName": "",
    "resourceType": ""
}
```
However, these keys will be defined like:

```json
{
    "Host": "",
    "Name": "",
    "Port": ""
}
```
So this would introduce an inconsistency in the data definitions. These keys should be formatted in lower camel-case to match the rest of the keys. If they need to displayed in title case in the HTML report, then jinja2 provides a filter called "title" to capitalize the first letter of the word: `{{ key | title }}`

This sort of formatting for display should be done in the jinja2 template instead of in the data itself. If keys do not title-case nicely, then it may just be worth hard-coding the table row header and then reference the expected data:

```
<table>
  <tr>
    <th>Host</th>
    <td>{{ info.host }}</td>
  </tr>
  <tr>
    <th>SSL Enabled</th>
    <td>{{ info.ssl }}</td>
  </tr>
  <!- continued for all rows -->
</table>
```

File: deployment_report/templates/viya_deployment_report.html.j2 [hunk 20]
```diff
@@ -130,6 +130,26 @@
   </div>
   {# Cluster Overview: Kubernetes Versions Accordion #}
 
+  {# Cluster Overview: Storage Accordion #}
+  {% set kind_name = "Storage" %}
+  <div class="jq-accordion">
+    <h2>Storage</h2>
+    <div>
+      {% for db, info in report_data.kubernetes.dbInfo.items() %}
+      <h7>Database({{ db }})</h7>
```
Comment (sasjowood) [inline deployment_report/templates/viya_deployment_report.html.j2:139]: Just a space needed between "Database" and the parenthetical: `<h7>Database ({{ db }})</h7>`.

**Review by sasjowood — CHANGES_REQUESTED** [review]

File: deployment_report/templates/viya_deployment_report.html.j2 [hunk 21]
```diff
@@ -136,11 +136,15 @@
...
     <div>
       {% for db, info in report_data.kubernetes.dbInfo.items() %}
-      <h7>Database({{ db }})</h7>
+      <h7>Database ({{ db }})</h7>
       <table>
       {% for key, value in info.items() %}
         <tr>
-          <th>{{ key }}</th>
+          {% if key == "sslEnabled" %}
+          <th>{{ "SSL_Enabled" }}</th>
```
Comment (sasjowood) [inline deployment_report/templates/viya_deployment_report.html.j2:144]: I think the underscore can be removed from SSL_Enabled since it's a table header: SSL Enabled

Other than this, I think things look good.

**Review by sasjowood — CHANGES_REQUESTED** [review]

**Review by sasjowood — APPROVED** [review]
Looks good. Thank you!
